File: leetcode/202.happy-number.go
// https://leetcode.cn/problems/happy-number/

func isHappy(n int) bool {
	m := map[int]bool{}
	for ; n != 1 && !m[n]; n, m[n] = sumOfSquares(n), true {}
	return n == 1
}

func sumOfSquares(n int) int {
	sum := 0
	for n > 0 {
		digit := n % 10
		sum += digit * digit
		n /= 10
	}
	return sum
}

File: leetcode/202.happy-number.py
# https://leetcode.cn/problems/happy-number/
class Solution:
    def isHappy(self, n: int) -> bool:
        return self.happySum(n, set({}))[1]

    def happySum(self, n: int, cache: set) -> tuple[int, bool]:
        if n in cache:
            return n, False
        else:
            cache.add(n)
        sum = 0
        for x in str(n):
            i = int(x)
            sum += (i ** 2)
        if sum != 1:
            return self.happySum(sum, cache)
        else:
            return sum, True
